Add SpatialJoinNode

mbasmanova opened the pull request:
Add a special node to represent spatial join to accommodate distributed spatial join #11072 that needs to add 3 more spatial join-specific properties (KDB tree and partition ID symbols for build and probe).

mbasmanova commented:
@kokosing Grzegorz, thank you for review. I'll work on your comments and update the PR.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SpatialJoinNode.java:
```diff
@@ -0,0 +1,124 @@
+     * must be from either left or right side of join. Symbols
+     * from left join side must precede symbols from right side
+     * of join.
+     */
```
@kokosing @sopel39 I copy-pasted this from `JoinNode`. I'm seeing that Karol added this to `JoinNode` in https://github.com/mbasmanova/presto/commit/2e9e968df23e6b9b1576e0fee9883c751231bc70 Karol, could you comment on this? Why do we need this in `JoinNode`?

kokosing replied:
This was recently removed from `JoinNode`, remove it from here also.

arhimondr replied:
I aggree. I made the same mistake when extending `TableWriterNode` for statstics. It is error prone. Because in case of symbols transformation you need to keep in mind to update this list as well.

mbasmanova replied:
@kokosing Grzegorz, I'm happy to remove this, but I don't see it removed from `JoinNode`. Is this change in the master?

kokosing replied:
The left comment in JoinNode is obsolete, it should be removed with https://github.com/prestodb/presto/commit/ea3e61ecbe970bbf248efc487cedd88db567da4f

mbasmanova replied:
@kokosing Thanks for the pointer and thanks for removing this comment from the `JoinNode`. I removed it from `SpatialJoinNode` as well.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+        @Override
+        public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
```
@kokosing `visitSpatialJoin` is a copy + delete of the `visitJoin`; I looked to see if I could reduce copy-paste, but I can't see how. Do you have any specific suggestions?

kokosing replied:
I thought about unifying `InnerJoinPushDownResult` and `OuterJoinPushDownResult`. They look very similar, that way we could reuse the code around the `switch`, but I am not sure if it is good refactor.

Let's leave it as it is.

arhimondr replied:
Is there any good reason why don't we want to have an hirerachy of joins?

Like: 

```
                     [JoinNode]
                     - type
                     - left
                     - rigth
                     - outputSymbols
           |             |
[EquiJoinNode]           [SpatialJoinNode]
- criteria                      - filter
- leftHashSymbol
- rightHashSymbol
- distributionType
- optional filter
```

(and maybe similar for `IndexJoinNode`)

This hirerarchy allows to remove most of the copy paste of the methods that do not return a new join node. Also it may help to remove some copy-paste for the rest of the methods.

CC: @kokosing @martint

kokosing replied:
I am ok to this. Let's save this refactor for another PR.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            // TODO: find a better way to directly optimize FALSE LITERAL in join predicate
+            if (newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL)) {
+                newJoinPredicate = new ComparisonExpression(ComparisonExpression.Operator.EQUAL, new LongLiteral("0"), new LongLiteral("1"));
```
why can't we leave `FALSE_LITERAL` here?

mbasmanova replied:
@kokosing I don't really know. This code was added in https://github.com/prestodb/presto/pull/4716/files to fix #4712 . @martint Martin, do you know why this check is needed? The fix added `AbstractTestQueries#testJoinWithInferredFalseJoinClause` test, so I tried commenting out this code and running the test. The test succeeded. Perhaps, this check is no longer needed?

martint replied:
I believe this is from the time when we didn't support non-equi joins.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                output = new FilterNode(idAllocator.getNextId(), output, postJoinPredicate);
+            }
+
+            output = new ProjectNode(idAllocator.getNextId(), output, Assignments.identity(node.getOutputSymbols()));
```
why is this extra identity projection needed?

mbasmanova replied:
@kokosing That's a good question. This logic was added in https://github.com/mbasmanova/presto/commit/84e9bab3302d043a3fe308668ff9016bba5daecf to fix #8673 and #8682 . I don't really understand why it is needed. @martint Martin, do you know? Would you explain? The tests added along the fix - `TestLogicalPlanner#testLeftConvertedToInnerInequalityJoinNoEquiJoinConjuncts` and `AbstractTestQueries#testNonEqualityLeftJoin` - seem to pass with and without this logic.

martint replied:
If I remember correctly, this was needed to preserve the outputs of the join node when a join with non-equi criteria was turned into a join with a filter on top and some underlying columns were only used by the filter. Probably something like:

```
- Join (on f(a)) :: [b, c]
  - t1
  - t2
```

=>

```
- Project :: [b, c]
  - Filter (f(a)) :: [a, b, c]
    - Join :: [a, b, c]
      - t1
      - t2
```

kokosing replied:
`JoinNode` contains a list of output symbols to preserve this information.

kokosing replied:
> but if the filter is pulled out 

Then the proper projection (with only join output symbols) should be added

mbasmanova replied:
@kokosing That's what I think this code is doing. Identity projections are created for output symbols of the *original* join node: `node.getOutputSymbols()`. However, none of this logic is needed for spatial joins. Hence, I took it out.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -409,6 +409,15 @@ public PlanOptimizers(
+                costCalculator,
+                ImmutableSet.<Rule<?>>builder()
+                        .add(new RemoveRedundantIdentityProjections())
+                        .addAll(new TransformSpatialPredicates(metadata).rules())
```
Why it was moved?

mbasmanova replied:
@kokosing To address the TODO which required a dedicated SpatialJoinNode. This type of rule logically belongs before `AddExchanges`, so I moved it.

```
        // TODO Make PredicatePushDown aware of spatial joins, move TransformSpatialPredicateToJoin	
        // before AddExchanges and update AddExchanges to set REPLICATED distribution for the build side.
```

mbasmanova commented:
@kokosing Added (copy-pasted) support for stats and cost.

mbasmanova commented:
@kokosing Grzegorz, would you take another look?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/cost/SpatialJoinStatsRule.java:
```diff
@@ -0,0 +1,74 @@
+
+        switch (node.getType()) {
+            case INNER:
+                return Optional.of(filterStatsCalculator.filterStats(crossJoinStats, node.getFilter(), session, types));
```
I'm working on removing the code that makes a wild `0.9` wild guess if nothing is know about the filters. But for now it will definitely produce something very wild, like `cross_product * 0.9`. But anyhow that is a exiting behaviour. Les't keep this class as is, so we can make sure that once we support spatial filter in filter estimates it will be enabled by default.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java:
```diff
@@ -275,6 +276,28 @@ public Expression visitSemiJoin(SemiJoinNode node, Void context)
+                            .addAll(pullNullableConjunctsThroughOuterJoin(extractConjuncts(rightPredicate), node.getOutputSymbols(), node.getRight().getOutputSymbols()::contains))
+                            .build());
+                default:
+                    throw new UnsupportedOperationException("Unknown spatial join type: " + node.getType());
```
nit: IllegalArgumentException, `Unsupported spatial join type: `

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformSpatialPredicates.java:
```diff
@@ -180,7 +181,7 @@ public Result apply(FilterNode node, Captures captures, Context context)
     public static final class TransformSpatialPredicateToLeftJoin
```
Yeah, i like `ExtractSpatialJoin` more. But i don't feel strong.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -395,6 +396,17 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, HashComputationSet pa
+        {
+            PlanWithProperties left = planAndEnforce(node.getLeft(), new HashComputationSet(), true, new HashComputationSet());
+            PlanWithProperties right = planAndEnforce(node.getRight(), new HashComputationSet(), true, new HashComputationSet());
+            checkState(left.getHashSymbols().isEmpty() && right.getHashSymbols().isEmpty());
```
ping

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
+            }
+
+            Verify.verify(postJoinPredicate.equals(TRUE_LITERAL), "Spatial join cannot have post join predicate");
```
static import.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+            Verify.verify(postJoinPredicate.equals(TRUE_LITERAL), "Spatial join cannot have post join predicate");
+
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            Verify.verify(!newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL), "Spatial join predicate is missing");
```
static import

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -437,6 +438,32 @@ public ActualProperties visitSemiJoin(SemiJoinNode node, List<ActualProperties>
+                    return ActualProperties.builderFrom(probeProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column)))
+                            .build();
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
```
`IllegalArgumentException`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -255,6 +256,26 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Set<Symbol>> con
+            Set<Symbol> seenSymbol = new HashSet<>();
+            List<Symbol> outputSymbols = node.getOutputSymbols().stream()
+                    .filter(context.get()::contains)
+                    .filter(seenSymbol::add)
```
You can use `distinct`

martint replied:
It's using `distinct()` right above the `filter(seenSymbol::add)`, which makes the call to `filter(seenSymbol::add)` and the `seenSymbol` set unnecessary. As a side note, try to avoid having side-effects when using streams (except for the final `forEach`). It makes it hard to read and reason about the code, and it's error prone.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java:
```diff
@@ -201,6 +202,20 @@ public StreamProperties visitJoin(JoinNode node, List<StreamProperties> inputPro
+                case LEFT:
+                    return leftProperties.translate(column -> PropertyDerivations.filterIfMissing(node.getOutputSymbols(), column));
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
```
`IllegalArgumentException`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SpatialJoinNode.java:
```diff
@@ -0,0 +1,151 @@
+                case LEFT:
+                    return Type.LEFT;
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + joinNodeType);
```
IllegalArgumentException

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SpatialJoinNode.java:
```diff
@@ -0,0 +1,151 @@
+    private final Expression filter;
+
+    @JsonCreator
+    public SpatialJoinNode(@JsonProperty("id") PlanNodeId id,
```
new line

arhimondr replied:
ping

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SpatialJoinNode.java:
```diff
@@ -0,0 +1,151 @@
+            @JsonProperty("filter") Expression filter)
+    {
+        super(id);
+        requireNonNull(filter, "filter is null");
```
remove

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java:
```diff
@@ -391,6 +392,42 @@ public Void visitSemiJoin(SemiJoinNode node, Set<Symbol> boundSymbols)
+                    predicateSymbols,
+                    allInputs);
+
+            int leftMaxPosition = -1;
```
This code isn't trivial. Please extract it to a method.

mbasmanova commented:
@kokosing @arhimondr Grzegorz, Andrii, thanks for your comments. I addressed them and added a commit to rename `TransformSpatialPredicates` to `ExtractSpatialJoins`.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java:
```diff
@@ -382,6 +371,47 @@ public Void visitSemiJoin(SemiJoinNode node, Set<Symbol> boundSymbols)
+            return null;
+        }
+
+        private void checkLeftOutputSymbolsBeforeRight(PlanNode left, List<Symbol> outputSymbols)
```
`private void checkLeftOutputSymbolsBeforeRight(List<Symbol> leftSymbols, List<Symbol> outputSymbols)`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                    newJoinPredicate = leftOuterJoinPushDownResult.getJoinPredicate();
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
```
IllegalArgumentException, use `UnsupportedOperationException` only if the entire method is not supported

arhimondr commented on presto-main/src/main/java/com/facebook/presto/cost/SpatialJoinStatsRule.java:
```diff
@@ -0,0 +1,74 @@
+            case LEFT:
+                return Optional.of(UNKNOWN_STATS);
+            default:
+                throw new IllegalStateException("Unknown spatial join type: " + node.getType());
```
IllegalArgumentException

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -607,6 +608,19 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, StreamPreferredProper
+            PlanWithProperties probe = planAndEnforce(
+                    node.getLeft(),
+                    defaultParallelism(session),
+                    parentPreferences.constrainTo(node.getLeft().getOutputSymbols()).withDefaultParallelism(session));
```
This is a little hard to read. I'd format as:

```java
parentPreferences.constrainTo(node.getLeft().getOutputSymbols())
                .withDefaultParallelism(session));
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+
+            switch (node.getType()) {
+                case INNER:
+                    InnerJoinPushDownResult innerJoinPushDownResult = processInnerJoin(inheritedPredicate,
```
First argument should go on the next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                    newJoinPredicate = innerJoinPushDownResult.getJoinPredicate();
+                    break;
+                case LEFT:
+                    OuterJoinPushDownResult leftOuterJoinPushDownResult = processLimitedOuterJoin(inheritedPredicate,
```
First argument should go on the next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                    throw new IllegalArgumentException("Unsupported spatial join type: " + node.getType());
+            }
+
+            verify(postJoinPredicate.equals(TRUE_LITERAL), "Spatial join cannot have post join predicate");
```
Why not? This would just be a filter on top of it, no?

martint replied:
Here's a query that should trigger that behavior for regular joins. You're going to need to adapt it for spatial joins:

```
select 1 
from orders o left join lineitem l on o.orderkey = l.orderkey 
where concat(o.comment, l.comment) is null
```

mbasmanova replied:
@martint Thanks. This query worked. I adapted to

```
select a.name, b.name
from points a left join polygons b on ST_Contains(ST_GeometryFromText(wkt), ST_Point(lng, lat))
where concat(a.name, b.name) is null
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+            verify(postJoinPredicate.equals(TRUE_LITERAL), "Spatial join cannot have post join predicate");
+
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            verify(!newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL), "Spatial join predicate is missing");
```
Can this happen through the normal course of expression simplification and optimization? In that case, it should turn into a regular join or even an empty ValuesNode.

mbasmanova replied:
@martint Maybe... I can't think of an example though and I'm reluctant to add code without being able to test it. What do you think?

martint replied:
It seems like it currently can't happen for anything other that full join due to lack of inference power. But it may start breaking later if we make the inference engine more capable.

mbasmanova replied:
@martint What would you recommend? Should I try to write this code and hope it works or leave it as is?

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -255,6 +256,27 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Set<Symbol>> con
+        @Override
+        public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Set<Symbol>> context)
+        {
+            Set<Symbol> expectedFilterInputs = ImmutableSet.<Symbol>builder()
```
This is misnamed. It's not just the inputs needed by the filter, but what the downstream node expects. I'd just call it `requiredInputs`.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:
```diff
@@ -53,21 +52,15 @@
     private final PlanNode left;
     private final PlanNode right;
     private final List<EquiJoinClause> criteria;
-    /**
```
Unrelated change? Move it to a separate commit.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:
```diff
@@ -53,21 +52,15 @@
 
     @JsonCreator
-    public JoinNode(@JsonProperty("id") PlanNodeId id,
+    public JoinNode(
```
Unrelated formatting change

mbasmanova commented:
@martint Martin, thank you for review. I made the changes you suggested. Would you take another look?
